### ISSUE TITLE
Cray PMI (libpals) support

### ISSUE THREAD

**Issue by jameshcorbett** [flux-framework/flux-coral2#6]
In order for Flux to support Cray's PMI on Shasta machines like rznevada, Perlmutter, and El Cap, Flux will need to interface with Cray's libpals. From the Cray folks:

> To support Cray PMI, the launcher must perform these tasks:
> 1. Assign an apid to the application (arbitrary string, unique per-application)
> 2. On each compute node, create a spool directory owned by the application's user
> 3. On each compute node, write an apinfo file in the spool directory (format below)
> 4. Set environment variables for each spawned process (listed below)
> 5. Remove the spool directory on each compute node when the application is complete
>  
> The apinfo file uses C structures written directly to the file. It starts with a header, followed by a number of structure arrays.
>  
> These environment variables should be set for each process:
> PALS_APID - Application ID (arbitrary string, mostly used for logging)
> PALS_APINFO - Full path to the apinfo file
> PALS_RANKID - Global rank ID for this process
> PALS_NODEID - Node index for this process (e.g. head compute node is 0, next compute node is 1, etc)
> PALS_SPOOL_DIR - Application-specific directory for keeping runtime files
> PMI_CONTROL_PORT - Port number for PMI to bind on each compute node. Must be different for each concurrent application running on the same node(s).

**Comment by dongahn** [comment]
@grondo: scheduling ports is an interesting idea. As is, though, we don't have an easy way to express "I want three port resources with the same numbers. We can say `node[2]->port[3]` but there will be no guarantee the three ports scheduled across two nodes will be identical. Were you thinking along this line? Or am I completely off?

**Comment by grondo** [comment]
I was thinking a "pool" of ports would be reserved at the cluster level. A broker module or something could bind to the configured port range at startup to guarantee availability. Each job would need one port (or more perhaps, but they would be the same across all nodes in the allocation), which each job shell would request from the local broker on startup and then hand off the job.

**Comment by dongahn** [comment]
> I was thinking a "pool" of ports would be reserved at the cluster level. 

I agree having some ports "reserved" for Flux use would make things simpler I think. 

> I was thinking a "pool" of ports would be reserved at the cluster level. 

Yeah I was thinking about that scheme as well. Where I got stuck was a case: we reserve K number of ports at the cluster level and if we have small jobs (e.g., 10K 1N jobs), we will need O(10K) ports while only one port number could do the job fine. 

We have have enough numbers of ephemeral ports that we can reserve so this might not be a problem. But maybe there could be a bit more resource efficient scheme.

**Comment by grondo** [comment]
>  But maybe there could be a bit more resource efficient scheme.

That would be nice. If we implemented the service to allocate a common port range as a broker service instead of shell plugin, maybe an initial small set of ports could be allocated (or inherited from the job) at instance startup. If these begin to run out because of the 10K jobs problem, then more could be allocated on demand and added to the pool.

This could help in nested instances as well if a large job ends up running many small jobs and it runs out of ports as well.

On the :coffee: call we also discussed a broker service that would simply hand out ports to local processes. If you run out, then you ask your parent instance (and up to the root).
